feat: add taxonomies for org api

rpenido opened the pull request:
## Description
This PR adds support for creating, viewing, updating and deleting Taxonomies for Organizations via REST APIs, with the permissions applied accordingly.

## Supporting information

* Depends on https://github.com/openedx/openedx-learning/pull/63
* Closes https://github.com/openedx/modular-learning/issues/74

## Testing instructions

* Check the unit tests in `test_views.py` and `test_rules.py` are correct.
* Run the unit tests

Manual check can be done in the CMS App:
- Run the cms app
- Login using Django Admin (non-staff users can login, they just can't see Django Admin)
  - http://127.0.0.1:8000/admin/
- REST API is here:
  - http://127.0.0.1:8000/api/content_tagging/v1/

---
Private-ref: [FAL-3450](https://tasks.opencraft.com/browse/FAL-3450)

rpenido commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,181 @@
+                    | Q(enabled=True)
+                )
+
+    filter_backends = [UserOrgFilterBackend]
```
I think this is the way to go. The whole rule/queryset problem is a common issue and is described here: https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0009-bp-permissions.html

rpenido replied:
I'm not sure about the code standards on this. If we think we should move this class to another place, let me know!

rpenido commented on openedx/features/content_tagging/rules.py:
```diff
@@ -88,7 +120,7 @@ def can_change_object_tag(user: User, object_tag: oel_tagging.ObjectTag = None)
 
 # Taxonomy
-rules.set_perm("oel_tagging.add_taxonomy", can_change_taxonomy)
+rules.set_perm("oel_tagging.add_taxonomy", can_add_taxonomy)
```
This was necessary to handle the difference in taxonomy=None from can_change_taxonomy.

rpenido commented on openedx/features/content_tagging/rules.py:
```diff
@@ -53,8 +84,9 @@ def can_change_taxonomy(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bo
-        not taxonomy or (taxonomy and not taxonomy.system_defined)
+
+    return not taxonomy or (
+        is_taxonomy_admin(user, taxonomy) and (not taxonomy.system_defined)
```
Everyone can call PUT/PATCH/DELETE (taxonomy=None). The actual object permission (taxonomy!=None) will be handled in the second call.

pomegranited replied:
Thank you for explaining this.. could you put that into the docstring for the method?

rpenido replied:
Done!

rpenido commented:
~~ToDo: Handle `{"ENABLE_CREATOR_GROUP": False}` when everything else is ok.~~
Done

pomegranited commented on cms/urls.py:
```diff
@@ -332,3 +332,8 @@
+
+# Content tagging
+urlpatterns += [
+    path('api/content_tagging/', include(('openedx.features.content_tagging.urls', 'content_tagging'), namespace='content_tagging')),
```
nit: do we need to specify both the `namespace` as a kwarg and in the `include`? 

Same question for `lms/urls.py` below.

rpenido replied:
Will check

rpenido replied:
This actually worked without any namespace. I tried to find a common pattern for URL but the file have a lot of different ones. :grimacing:

pomegranited commented on openedx/features/content_tagging/models.py:
```diff
@@ -67,19 +70,21 @@ def get_relationships(
             rel_type=rel_type,
         )
+        if taxonomy:
+            rels = rels.filter(taxonomy=taxonomy)
```
I'm not clear why this change is needed.. can you explain which use case it supports?

rpenido replied:
Thank you for this @pomegranited. We will not have taxonomy=None here.

pomegranited commented on openedx/features/content_tagging/models.py:
```diff
@@ -67,19 +70,21 @@ def get_relationships(
+            rels = rels.filter(taxonomy=taxonomy)
         # A relationship with org=None means all Organizations
-        if rels.filter(org=None).exists():
+        if (not taxonomy or taxonomy.enabled) and rels.filter(org=None).exists():
```
I'm uncomfortable with having [`rules` logic](https://github.com/openedx/edx-platform/pull/32871/files#diff-eb87bf7bd1876f196e881debb28be1ef1614e44faad929f0ccc111044cf9954fR75) copied in here..

Could you point me to the use case that requires this change?

rpenido replied:
An enabled=True taxonomy from "orgA" can be viewed by a user in any org (i.e. "orgB"). If we change this (and I think we should), this code will go. We will also get rid off some other enabled=True filters in the code.

pomegranited replied:
Beautiful, yes I agree that's how it should work.
I've updated https://github.com/openedx/modular-learning/issues/74 to reflect this, so now: `userA` never sees `tB*`, and `userB` never sees `tA*`.

rpenido replied:
Just to document: this is not valid anymore. We went for simpler permissions.

pomegranited commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,184 @@
+
+class TaxonomyOrgView(TaxonomyView):
+    """
+    ToDo: Fix the docstring
```
It looks pretty "fixed" to me.. 😄 What else needs doing here?

pomegranited commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,184 @@
+                            )
+                        )
+                    )
+                    | Q(enabled=True)
```
I don't know whether this detailed model filtering logic should live in `content_tagging.models` or `content_tagging.api`, but it shouldn't live here.. 

Plus:
* This filter misses where the user is not a global taxonomy admin, but the TaxonomyOrg has `org=None` (instance taxonomy).
  To keep that logic in one place, could we maybe modify `TaxonomyOrg.get_relationships()` to take a list of Organizations instead of just a single `org_short_name`, and use that result in the `Exists` clause?
* the `enabled=True` check is far faster than the org checks, and so should be listed first.

rpenido replied:
> * This filter misses where the user is not a global taxonomy admin, but the TaxonomyOrg has `org=None` (instance taxonomy).

The `enabled=True` handle these cases (because we don't want to show `enabled=False, org=None` to org. But this code changed because we don't want a user from 'orgA' to list taxonomies from 'orgB'

```python
   def filter_queryset(self, request, queryset, view):
            if is_global_taxonomy_admin(request.user):
                return queryset
            else:
                user_orgs = get_user_taxonomy_orgs(request.user)
                user_orgs_short_name = [org.short_name for org in user_orgs]
                return queryset.filter(
                    (
                        Q(enabled=True)
                        & Q(
                            Exists(
                                TaxonomyOrg.objects.filter(
                                    taxonomy=OuterRef("pk"),
                                    rel_type=TaxonomyOrg.RelType.OWNER,
                                    org=None,
                                )
                            )
                        )
                    )
                    | Q(
                        Exists(
                            TaxonomyOrg.objects.filter(
                                taxonomy=OuterRef("pk"),
                                rel_type=TaxonomyOrg.RelType.OWNER,
                                org__short_name__in=user_orgs_short_name,
                            )
                        )
                    )
                )
``

rpenido replied:
> I don't know whether this detailed model filtering logic should live in `content_tagging.models` or `content_tagging.api`, but it shouldn't live here..

I'm not sure if I agree with you on that.

Currently, the API doesn't handle the permissions or "know" the rules. The rules are secured on the ViewSet via `TaxonomyObjectPermissions`. Similarly, I think the 'permission filters' should be part of the view (via  `UserOrgFilterBackend`). 

I also thought about using the `UserOrgFilterBackend` but leaving the filter structure (the `Q` query) in the model. Still, that way, we will inject permission code in the model (and I don't think this is good either).

If you think both (rules and filters) should be the API's responsibility, I'm also good with it. We could call the has_perm inside the API. This will lead to more code because the Django View has a lot of mixins to handle that automagically.

If you are worried about bloating the View code (and I think it is now), I will move this class to a `filters.py` (but let the view handle the permission). Then, we have a pair of `rules.py` and `filters.py` for what a user can do and what he is allowed to view.

Do you think this makes sense? Are there other concerns that I'm not getting here?

pomegranited replied:
@rpenido I'd really hoped that all the fancy model filtering logic could live in TaxonomyOrg, since that's what understands its internal relationships.. but see what shakes out with [this change](https://github.com/openedx/edx-platform/pull/32871#discussion_r1281250100).

> If you are worried about bloating the View code (and I think it is now), I will move this class to a filters.py (but let the view handle the permission). Then, we have a pair of rules.py and filters.py for what a user can do and what he is allowed to view.

👍

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -87,15 +87,35 @@ def _expected_users_have_perm(
+            self.user_both_orgs.has_perm(perm) == learner_perm
+            if perm not in ("oel_tagging.add_taxonomy", "oel_tagging.change_taxonomy")
+            else True
+        )
```
So.. I'm not sure, but I think this is the same thing as saying:

```python
        if perm not in ("oel_tagging.add_taxonomy", "oel_tagging.change_taxonomy"):
            assert (self.user_both_orgs.has_perm(perm) == learner_perm
        else:    
            assert(True)
```

Which would mean that we're not actually checking `self.user_both_orgs.has_perm("oel_tagging.add_taxonomy")` or `self.user_both_orgs.has_perm("oel_tagging.change_taxonomy")` at all. And we should be.

I could be wrong, but either way, it's clearer if we split out all the `if perm in ("oel_tagging.add_taxonomy", "oel_tagging.change_taxonomy")` asserts together, and the `else` asserts in another block.

I'm also wincing because this is pretty awkward.. but I created one single `_expected_users_have_perm` method that everything uses, so that awkwardness is my doing..

rpenido replied:
> I could be wrong, but either way, it's clearer if we split out all the `if perm in ("oel_tagging.add_taxonomy", "oel_tagging.change_taxonomy")` asserts together, and the `else` asserts in another block.

Agree! That way the code will be cleaner

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -156,7 +174,9 @@ def test_view_taxonomy_enabled(self, enabled, taxonomy_attr):
-        self._expected_users_have_perm(perm, taxonomy, learner_obj=enabled)
+        self._expected_users_have_perm(
+            perm, taxonomy, learner_perm=True, learner_obj=enabled
+        )
```
Could you please set your max line length to 120 characters and reformat to avoid these changes? cf [EdX Python Style Guide](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#syntax-and-organization)

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -472,10 +492,14 @@ def test_no_orgs_no_perms(self, perm, taxonomy_attr):
         assert self.learner.has_perm(perm)
-        assert not self.learner.has_perm(perm, taxonomy)
+        if perm != "oel_tagging.add_taxonomy":
+            assert not self.learner.has_perm(perm, taxonomy)
```
This change means we're not testing permissions at all when `perm == "oel_tagging.add_taxonomy"`, and we need to.

Also, if the checks for that permission differ from `change_taxonomy` and `delete_taxonomy`, then it's better to split `test_no_orgs_no_perms` into two test methods, one for the `ddt.data("oel_tagging.add_taxonomy")` and one for the other checks.

rpenido replied:
> This change means we're not testing permissions at all when `perm == "oel_tagging.add_taxonomy"`, and we need to.

This is intended. For "oel_tagging.add_taxonomy", calling `has_perm(perm)` make sense (and we are calling it), but calling `has_perm(perm, taxonomy)` doesn't. We don't have a taxonomy object to test permission on.

I will split the tests as you suggested!

pomegranited commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,184 @@
+                        )
+                    )
+                    | Q(enabled=True)
+                )
```
I don't know whether this detailed model filtering logic should live in `content_tagging.models` or `content_tagging.api`, but it shouldn't live here.. I'd rather it looked similar to your `get_queryset` below.

Also note:
* The `Exists` clause here misses where the user is not a global taxonomy admin, but the TaxonomyOrg has `org=None` (instance taxonomy).
* the `enabled=True` check is far faster than the org checks, and so should be listed first.

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -1,16 +1,32 @@
+    return False
+
+
 def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool:
```
Can you use `is_global_taxonomy_admin` to remove the duplicated logic in `is_taxonomy_admin`?

e.g.

    if is_global_taxonomy_admin(user):
        return True

    taxonomy_orgs = TaxonomyOrg.get_organizations...
```

rpenido replied:
Done!

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -43,7 +71,16 @@ def can_view_taxonomy(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
         taxonomy = taxonomy.cast()
-    return (taxonomy and taxonomy.enabled) or is_taxonomy_admin(user, taxonomy)
+
+    return not taxonomy or taxonomy.enabled or is_taxonomy_admin(user, taxonomy)
```
Could you update the docstring here to explain 

```
Anyone can view a list of taxonomies (i.e. taxonomy==None) or an enabled taxonomy,
but only taxonomy admins can view a disabled taxonomy.
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -43,7 +71,16 @@ def can_view_taxonomy(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+@rules.predicate
+def can_add_taxonomy(user: User) -> bool:
+    """
+    Even taxonomy admins cannot change system taxonomies.
```
Please update the docstring here.

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+    assert data["allow_multiple"] == taxonomy.allow_multiple
+    assert data["allow_free_text"] == taxonomy.allow_free_text
+    assert data["system_defined"] == taxonomy.system_defined
+    assert data["visible_to_authors"] == taxonomy.visible_to_authors
```
There's a few ways to convert a model instance to a dict, so you could collapse these two methods: https://stackoverflow.com/a/29088221

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+        )
+        # Orphaned taxonomy
+        # self.t3 = Taxonomy.objects.create(name="t3", enabled=True)
+        # self.t4 = Taxonomy.objects.create(name="t4", enabled=False)
```
Yes -- it's worth checking that the Taxonomies with no orgs also behave as expected, i.e. that only global taxonomy admins can edit them, but anyone can view the enabled one?

rpenido replied:
Should we add this to the Issue description?
The code and the tests already cove this.

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+
+@ddt.ddt
+@override_settings(FEATURES={"ENABLE_CREATOR_GROUP": True})
+class TestTaxonomyViewSet(TestTaxonomyViewSetMixin, APITestCase):
```
These test cases make me very happy!

pomegranited replied:
This isn't your fault @rpenido , but this change really highlights how writing tests to match the acceptance criteria (as I suggested you do) makes the tests themselves illegible.

I've raised https://github.com/openedx/modular-learning/issues/89 as a separate issue to refactor these tests (and `test_rules.py`) to remove most of the `ddt.data` confusion, and make the tests more legible, as @bradenmacdonald  has suggested.

rpenido replied:
Copying what i wrote below:

@pomegranited Just my 2 cents here: When my brain was fried trying to figure out which permissions were wrong, I found it easier to deal with a large dtt function than a helper function.

The helper function will stop in the first assert failure. Sometimes I fixed the first condition only to discover that the second condition also failed and got trapped, changing the behavior without a clear view of all failing conditions.

The ddt always showed me all mismatched permissions.

Maybe I was just more comfortable with ddt because I wrote it (I'm not sure).

I just want to give you this insight. I'm following your lead here and don't disagree with your approach.

There's a discussion about this somewhere (besides the issue you created)? I'm interested in the results and want to follow.

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+
+    @ddt.data(
+        (None, status.HTTP_403_FORBIDDEN),
+        ("userA", status.HTTP_403_FORBIDDEN),
```
Could you also ensure `user` gets `status.HTTP_403_FORBIDDEN`, just for completeness?

rpenido replied:
Done!

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+        ("userS", "tB2", status.HTTP_204_NO_CONTENT),
+        ("userS", "tC2", status.HTTP_204_NO_CONTENT),
+        ("userS", "t1", status.HTTP_204_NO_CONTENT),
+        ("userS", "t2", status.HTTP_204_NO_CONTENT),
```
Can you add checks for deleting the system taxonomy too, just in case?
```suggestion
        ("userS", "t2", status.HTTP_204_NO_CONTENT),
        ("userS", "st1", status.HTTP_403_FORBIDDEN),
        ("userS", "st2", status.HTTP_403_FORBIDDEN),
```

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+            "user",
+            None,
+            None,
+            ("st1", "st2", "t1", "t2", "tA1", "tA2", "tB1", "tB2", "tC1", "tC2"),
```
Even if the creator group is disabled, only `userS` should be allowed to see disabled taxonomies in the list, or in the detail view.

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+        (None, "t1", status.HTTP_403_FORBIDDEN),
+        ("user", "t1", status.HTTP_200_OK),
+        ("user", "tA1", status.HTTP_200_OK),
+        ("user", "tA2", status.HTTP_200_OK),
```
Same here.. I think only `userS` should be allowed to update/patch/delete disabled taxonomies..

*Though I get why that's weird, since I think we let users *create* disabled taxonomies.. But we can tune this when we have a UI to support.)

rpenido replied:
I blocked the users to update/patch/delete global taxonomies (org=None). But in this case, every user has CourseCreatorRole, so is hard to differentiate the permission from other users (user vs userA).

pomegranited replied:
Ah ok -- this is with the content creator groups disabled, so everyone is an editor. That makes sense.

pomegranited reviewed (changes requested):
Looking good @rpenido , especially the tests!
Let me know if you have any questions or objections to the changes I've suggested?

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,964 @@
+        (None, "t1", status.HTTP_403_FORBIDDEN),
+        ("user", "t1", status.HTTP_403_FORBIDDEN),
+        ("user", "tA1", status.HTTP_403_FORBIDDEN),
+        ("userA", "tA1", status.HTTP_200_OK),
```
Can you add more checks for the instance-wide taxonomies too?  `userA` should be forbidden from updating/patching `t1` and `t2`.

rpenido replied:
Done

pomegranited commented on requirements/edx/kernel.in:
```diff
@@ -115,7 +115,10 @@ openedx-calc                        # Library supporting mathematical calculatio
-openedx-learning<=0.1
+# FIXME Use pip package once created
+# ref https://github.com/openedx/openedx-learning/pull/62
+# need new release
```
Yeah sorry -- we need to bump the version and tag a new release to do this, cf https://github.com/openedx/openedx-learning/pull/67#issuecomment-1659802119

pomegranited commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,155 @@
+        page_size = 100
+        max_page_size = 1000
+
+    pagination_class = TaxonomyPagination
```
I'm sorry, didn't notice that the pagination was only added here instead of on the openedx-learning side..
Do you mind to create a small PR against openedx-learning that moves this over there?

rpenido replied:
Done: https://github.com/openedx/openedx-learning/pull/69

rpenido replied:
ToDo: remove pagination when the PR is merged:
* https://github.com/openedx/openedx-learning/pull/69

pomegranited replied:
Your PR is merged and now available with `openedx-learning==0.1.2`.

If you run `make requirements` in your LMS shell, that should bump the installed version in the `requirements/edx/*` files.
And if you run into this issue, see [these comments on stackoverflow](https://stackoverflow.com/q/76585758):
```
Exception: Can not find valid pkg-config name.
      Specify MYSQLCLIENT_CFLAGS and MYSQLCLIENT_LDFLAGS env vars manually
```

pomegranited commented on openedx/features/content_tagging/models.py:
```diff
@@ -73,13 +76,14 @@ def get_organizations(
             rel_type=rel_type,
+            taxonomy=taxonomy
         )
+        # ToDo: Remove. We don't want this here. We want to retur only direct orgs relations.
```
note to self: check this.

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -60,23 +64,45 @@ def setUp(self):
-        Checks that all users have the given permission on the given object.
-
-        If learners_too, then the learner user should have it too.
+        Checks that all users have the given permission on the given taxonomy.
```
nit: this isn't just about taxonomy perms:
```suggestion
        Checks that all users have the given permission on the given object.
```

rpenido replied:
Reverted this change

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -87,37 +113,58 @@ def _expected_users_have_perm(
 
         # Learners can't do much but view
-        assert self.learner.has_perm(perm) == learner_perm
+        assert self.learner.has_perm(perm)
```
Please ensure the parameters to this helper method are used as expected -- `learner_perm` should be used when checking `self.learner`, just like `learner_obj` is used to check the learner's object permissions. If the param is no longer needed, or has a new purpose, then it's ok to remove or rename.

rpenido replied:
Reverted this change and created a new helper function

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -87,37 +113,58 @@ def _expected_users_have_perm(
+
+        # Org content creators can't create new taxonomies
+        assert self.user_both_orgs.has_perm(perm) == learner_perm
+        assert self.user_org2.has_perm(perm) == learner_perm
```
Ditto here -- `learner_perm` is being used for something else.

rpenido replied:
Reverted

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,72 +45,134 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
-    Anyone can view an enabled taxonomy,
-    but only taxonomy admins can view a disabled taxonomy.
+    Everyone can potentially view a taxonomy (taxonomy=None). The object permission will be checked on a
+    second pass to determine if the user can change this specific taxonomy.
```
```suggestion
    Everyone can potentially view a taxonomy (taxonomy=None). The object permission must be checked
    to determine if the user can view a specific taxonomy.
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,72 +45,134 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
-    but only taxonomy admins can view a disabled taxonomy.
+    Everyone can potentially view a taxonomy (taxonomy=None). The object permission will be checked on a
+    second pass to determine if the user can change this specific taxonomy.
+    Only taxonomy admins can view a disabled taxonomy. #ToDo: Check if this is correct.
```
That's correct.

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,72 +45,134 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
 def can_change_taxonomy(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool:
     """
+    Everyone can potentially change a taxonomy (taxonomy=None). The object permission will be checked on a
+    second pass to determine if the user can change this specific taxonomy.
```
nit:
```suggestion
    Everyone can potentially change a taxonomy (taxonomy=None). The object permission must be checked
    to determine if the user can change a specific taxonomy.
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,72 +45,134 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
     """
-    Even taxonomy admins cannot add tags to system taxonomies (their tags are system-defined), or free-text taxonomies
+    Everyone can potentially change a taxonomy tag (tag=None). The object permission will be checked on a
+    second pass to determine if the user can change this specific tag.
```
```suggestion
    Everyone can potentially change a taxonomy tag (tag=None). The object permission must be checked
    to determine if the user can change a specific tag.
```

rpenido replied:
Done!

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,72 +45,134 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    """
+    Only taxonomy admins can add taxonomy tags.
+    """
+    return is_taxonomy_admin(user)
```
This is not true.. Normal content authors must to be able to create object tags (i.e tag their content).

rpenido replied:
Reverted these changes

openedx-webhooks commented:
Thanks for the pull request, @rpenido!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

rpenido commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -101,37 +101,88 @@ def _expected_users_have_perm(
     @ddt.data(
-        ("oel_tagging.add_taxonomy", "taxonomy_all_orgs"),
-        ("oel_tagging.add_taxonomy", "taxonomy_both_orgs"),
-        ("oel_tagging.add_taxonomy", "taxonomy_disabled"),
```
This code was checking `add_taxonomy` against an existing object.

rpenido commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -101,37 +101,88 @@ def _expected_users_have_perm(
+    )
+    def test_can_change_taxonomy(self, perm):
+        # Everyone has the permission not related to object
+        assert self.superuser.has_perm(perm)
```
Test permissions not related to the object one time

rpenido commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -446,11 +528,6 @@ def test_object_tag_no_taxonomy(self, perm):
-        ("oel_tagging.add_taxonomy", "taxonomy_both_orgs"),
-        ("oel_tagging.add_taxonomy", "taxonomy_disabled"),
-        ("oel_tagging.add_taxonomy", "taxonomy_one_org"),
-        ("oel_tagging.add_taxonomy", "taxonomy_no_orgs"),
```
These tests are done in the base class.

rpenido commented:
Hi @pomegranited! I reverted the changes that are not related to the Taxonomy model.

pomegranited commented on openedx/features/content_tagging/models/base.py:
```diff
@@ -81,6 +82,17 @@ def get_organizations(
+            taxonomy=taxonomy,
+            rel_type=rel_type,
+        )
+        return [rel.org for rel in rels if rel.org is not None]
```
You can make this more efficient by excluding the `null` orgs at the query level, and also select the `org` value in the join (instead of as a separate query per relationship):

```suggestion
        ).exclude(org=None).select_related('org')
        return [rel.org for rel in rels]
```

rpenido replied:
Thank you! Done efd97f5d5457ae9e18012751064dfdaddfd53930

pomegranited commented on openedx/features/content_tagging/rest_api/v1/filters.py:
```diff
@@ -0,0 +1,45 @@
+        else:
+            user_orgs = get_user_taxonomy_orgs(request.user)
+            user_orgs_short_name = [org.short_name for org in user_orgs]
+            return queryset.filter(
```
nit: we can avoid indenting the rest of this method under the `else` case because your `if` statement above returns out of the function, e.g.

```suggestion
        user_orgs = get_user_taxonomy_orgs(request.user)
        user_orgs_short_name = [org.short_name for org in user_orgs]
        return queryset.filter(
        ...
```

rpenido replied:
Done 85df8e1b0dfdf477c13a6db8a4ea9f267c838a40

pomegranited commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,156 @@
+    View to list, create, retrieve, update, or delete Taxonomies.
+
+        NOTE: Only the GET request requires a request parameter, otherwise pass the uuid as part
+        of the post body
```
Is this comment correct? We need the `request` parameter to identify the logged-in user, and isn't it required for all of these endpoints?

Also, a `uuid` is a specific thing: https://docs.python.org/3/library/uuid.html
So saying `taxonomy id` would be more appropriate here.

But I think it's better to remove this line entirely; users of the API can consult the detailed docs for each endpoint below.

rpenido replied:
Sorry. This line doesn't make sense at all.
Fixed here: 8fb730c98ec9083fe144c05668be2d9acd36f228

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +36,108 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    """
+    Returns True if the given user is a Global Taxonomy Admin or a course creator who can create courses for any org.
+    """
+    return is_global_taxonomy_admin(user) or is_content_creator(user, None)
```
This method isn't used anywhere, so can be removed:
```suggestion
```

rpenido replied:
Done c57e9d3fc3a1df9e373c98764f5697e28270bdaf

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +36,108 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    Global Taxonomy Admins include global staff and superusers.
+    """
+    if oel_tagging.is_taxonomy_admin(user):
+        return True
```
This method returns `None` implicitly if the user is not a taxonomy admin.. could you write it like this instead?
```suggestion
    return oel_tagging.is_taxonomy_admin(user)
```

rpenido replied:
Done 067c5bee849de9c57454387e66c14e20b054ddd7

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +36,108 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+
+    taxonomy = taxonomy.cast()
+
+    return is_taxonomy_admin(user, taxonomy) or taxonomy.enabled
```
Performance-wise, it's cheaper to check `taxonomy.enabled`, so could you check that first?

```suggestion
    return taxonomy.enabled or is_taxonomy_admin(user, taxonomy)
```

rpenido replied:
Done: 1541d0daee93dadca3168780122cf818f7e83353
I will pay more attention to that!

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +36,108 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    """
+    Only global taxonomy admins add taxonomies.
+    There is no API method to create a taxonomy associated with an organization
+    neither to add a taxonomy to an organization.
```
So, this is true, but it may change.. could you phrase it a little softer?
```suggestion
    Only global taxonomy admins can add taxonomies (for now).
    
    To do: There is currently no REST API method to create a taxonomy associated with an organization
    neither to add a taxonomy to an organization.
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +36,108 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+
+    taxonomy = taxonomy.cast()
+
+    return is_taxonomy_admin(user, taxonomy) and (not taxonomy.system_defined)
```
Similar performance nit:
```suggestion
    return (not taxonomy.system_defined) and is_taxonomy_admin(user, taxonomy)
```

pomegranited replied:
Also, your change doesn't actually change the logic of this method. It does clarify it though, so I'm ok with keeping it. But in future, could you please avoid making unnecessary changes to keep the diff as small as possible?

rpenido replied:
Sure! I remove some logic redundancies overall in the checks, but I agree with you. I will try to keep the changes small next time!

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -1,24 +1,25 @@
 
 
-def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool:
+def is_taxonomy_user(user: User, taxonomy: oel_tagging.Taxonomy) -> bool:
```
This method does sometimes take `taxonomy=None` from the checks below, so could you please re-add that?
```suggestion
def is_taxonomy_user(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool:
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -1,24 +1,25 @@
-    Global Taxonomy Admins include global staff and superusers, plus course creators who can create courses for any org.
-    Otherwise, a taxonomy must be provided to determine if the user is a org-level course creator for one of the
-    taxonomy's org owners.
+    Returns True if the given user is a Taxonomy User for the given content taxonomy.
```
Please explain more:
```suggestion
    Returns True if the given user is a Taxonomy User for the given content taxonomy.
    
    Taxonomy users include global staff and superusers, plus course creators who can create courses for any org.
    Otherwise, we need a taxonomy provided to determine if the user is an org-level course creator for one of the orgs allowed to use this taxonomy.
```

pomegranited commented on openedx/features/content_tagging/rest_api/v1/filters.py:
```diff
@@ -0,0 +1,45 @@
+            return queryset
+        else:
+            user_orgs = get_user_taxonomy_orgs(request.user)
+            user_orgs_short_name = [org.short_name for org in user_orgs]
```
Hmm.. since `get_user_taxonomy_orgs` is only used here, and half of that method's logic is in the lines above with your `is_global_taxonomy_admin` check.. I think this should be refactored.

Maybe replace `rules.get_user_taxonomy_orgs` with this, so we're still isolating the `is_content_creator` check inside the rules:

```
def is_taxonomy_admin_for_org(user: User, org: Organization) -> bool:
    return is_content_creator(user, org.short_name)
```

And do your org filtering here:
```suggestion
            user_orgs_short_name = [
                org.short_name if is_taxonomy_admin_for_org(user, org)
                for org in Organization.objects.all()
            ]
```

rpenido replied:
Done 98cb2eb68472c30b201988a2e5aa7a041104da98

pomegranited commented on openedx/features/content_tagging/tests/test_rules.py:
```diff
@@ -101,37 +101,88 @@ def _expected_users_have_perm(
         assert self.learner.has_perm(perm) == learner_perm
         assert self.learner.has_perm(perm, obj) == learner_obj
 
+    def _expected_users_have_perm_taxonomy(
```
As noted above, if I'd written these tests more legibly to begin with, we wouldn't have to do this here. Will address as part of https://github.com/openedx/modular-learning/issues/89.

pomegranited reviewed (changes requested):
Hi @rpenido , I've requested a few changes, and noted some things that I'd like to fix as part of a separate task. 

I'm also awaiting word from Product on how they want the org content author permissions to work, something that came up in our last meeting. So this might sit in External Review for a little while before it can be merged.

But if you could address the requested changes, I'll work on the rest and let you know.

Thank you!

rpenido commented:
Hi @pomegranited!

I made the changes and also reverted some "style only" modifications to get a clear diff. Thank you for your input on that.

I will squash/rebase when we are ready to External Review!

rpenido commented:
Hi @pomegranited !
The change in permissions requested is here: b10b02a2288145408c9e6614bdf0d719c1f37f3f (including fixing the view tests and cleaning some code)
I cherry-picked the new `test_rules.py` and fixed it with the new conditions here: 3dccf60435e94d9b4ddf16d2f6b300866a690269

~~- [ ] ToDo: change the test_views to use helper functions
Is the proposal to make a helper function that checks list, get, edit, update, delete (for taxonomy, tag, objecttag), and then duplicate this function for each "user role"?
We will end up with multiple methods and views in the same test. I'm not sure about this one in the `test_view` context. Do you think this is the way to go?~~

pomegranited commented on openedx/features/content_tagging/rest_api/v1/filters.py:
```diff
@@ -0,0 +1,45 @@
+                    )
+                )
+            )
+        )
```
We only need the `enabled` filter here -- the rest of this org checking can go.

```suggestion
        return queryset.filter(enabled=True)
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +38,61 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    Taxonomy users include global staff and superusers, course creators who can create courses for any org and
+    content creators for this specific org.
+    """
+    return is_content_creator(user, org.short_name)
```
As noted, we don't need any of this "is_content_creator" logic anymore. Only global staff + superusers can add/edit taxonomies, and anyone can view an enabled taxonomy, no matter its org.

rpenido replied:
Done!

rpenido commented:
@pomegranited Permissions fixed, tests adjusted, squashed, and rebased!

I made a minor refactor in `test_rules.py`, so reviewing the file instead of the diff will be better.

Let me know if I need to include something!

pomegranited commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,153 @@
+    View to list, create, retrieve, update, or delete Taxonomies.
+
+    **List Query Parameters**
+        * enabled (optional) - Filter by enabled status. Valid values: true, false, 1, 0, "true", "false", "1"
```
It's worth noting that this `enabled` parameter is only useful for Taxonomy Admins, since everyone else can only see taxonomies where `enabled=True`.

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -67,7 +93,7 @@ def can_change_taxonomy_tag(user: User, tag: oel_tagging.Tag = None) -> bool:
     if taxonomy:
         taxonomy = taxonomy.cast()
-    return is_taxonomy_admin(user, taxonomy) and (
+    return is_taxonomy_user(user, taxonomy) and (
```
Please revert this change -- only Taxonomy Admins can change taxonomy tags.

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -82,13 +108,13 @@ def can_change_object_tag(user: User, object_tag: oel_tagging.ObjectTag = None)
     taxonomy = object_tag.taxonomy if object_tag else None
```
Could you update the docstring for this method, since its rules have changed?

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -11,14 +11,16 @@
+    orgs allowed to use this taxonomy.
     """
+    # ToDo: This is a temporary fix to not break can_change_taxonomy_tag and can_change_object_tag
+    # Should be revised when the API that uses these roles is implemented
```
Could you remove these lines? We don't need to excuse this method anymore; it's used by `can_change_object_tag`, and that's fine.
```suggestion
```

pomegranited commented on openedx/features/content_tagging/rules.py:
```diff
@@ -35,27 +37,51 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    Only taxonomy admins can add taxonomies.
+
+    To do: There is currently no REST API method to create a taxonomy associated with an organization
+    neither to add a taxonomy to an organization.
```
Please remove these lines -- These REST API methods are part of implementing the frontend requirements, and so we won't forget to add them.

```suggestion
```

pomegranited reviewed (changes requested):
We're getting there! Couple of comment-related changes, and one functionality change to revert.

Thank you @rpenido , this is looking great.

rpenido commented:
Thank you for the review @pomegranited !
I updated the branch.

Let me know if we need something else to get this to external review.

pomegranited commented on openedx/features/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -0,0 +1,637 @@
+    @ddt.data(
+        ("user", None, None, ("ot1", "st1", "t1", "tA1", "tB1", "tC1")),
+        ("userA", None, None, ("ot1", "st1", "t1", "tA1", "tB1", "tC1")),
+        ("userS", None, None, ("ot1", "ot2", "st1", "st2", "t1", "t2", "tA1", "tA2", "tB1", "tB2", "tC1", "tC2")),
```
Since the new default pagination has a `page_size = 10`, we don't get the final `tC1` and `tC2` taxonomies in this list, so this test fails (in both TestCase classes).

It's ok to just remove these  -- when we specify `orgA` below, we see that these taxonomies are returned.

```suggestion
        # Default page_size=10, and so "tC1", "tC2" appear on page 2.
        ("userS", None, None, ("ot1", "ot2", "st1", "st2", "t1", "t2", "tA1", "tA2", "tB1", "tB2")),
```

rpenido replied:
Thank you @pomegranited ! I missed this failed test.
This is fixed now!

pomegranited reviewed (approved):
:+1: This looks good, thank you @rpenido !

I'm seeing one issue when I run the tests (noted below), but it's a quick fix. And I merged latest `master` to resolve the conflicts in requirements -- hopefully that allows the CI tests to run again?

@bradenmacdonald  this is ready for your review when you get a chance.

- [x] I tested this using the PR instructions to login as a staff user and use the taxonomy REST API. 
- [x] I read through the code and tests to ensure they met the permissions requirements.
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation for the REST API
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

bradenmacdonald commented on openedx/features/content_tagging/rest_api/v1/views.py:
```diff
@@ -0,0 +1,147 @@
+
+class TaxonomyOrgView(TaxonomyView):
+    """
+    View to list, create, retrieve, update, or delete Taxonomies.
```
This is just the description of `TaxonomyView`. Please explain how/why this subclass changes `TaxonomyView`.

Also, I'm not sure it's a good idea to duplicate the entire docstring of `TaxonomyView` here, because we have to manually keep them in sync now. What do you think about just saying, "refer to TaxonomyView docstring for usage details" ?

rpenido replied:
> This is just the description of `TaxonomyView`. Please explain how/why this subclass changes `TaxonomyView`.

Done f39182d6b92a3fd96762472e8e9429da262d5b01. I also fixed the URL in the docstring.

> Also, I'm not sure it's a good idea to duplicate the entire docstring of `TaxonomyView` here, because we have to manually keep them in sync now. What do you think about just saying, "refer to TaxonomyView docstring for usage details" ?

I didn't like duplicating it here also, but this text is used in the swagger docs of the API. Having all the info available while consuming the API is also important. I'm not sure which is the best balance here.

![image](https://github.com/openedx/edx-platform/assets/849463/822b9248-8375-4007-8062-3defaa66b8f5)

bradenmacdonald replied:
I see. Maybe you can [set the docstring programmatically](https://stackoverflow.com/questions/4056983/how-do-i-programmatically-set-the-docstring) so it's not duplicated in the code, but still available in the swagger UI? But if that's too messy, don't worry about it.

rpenido replied:
I liked the solution, but looking at our codebase, I didn't find any use of it. :worried: Maybe it is too much?

Wrong (out sync) documentation is worse than less documentation, so I opted for the "refer to.." as you suggested in the first place.
Commit: 7ab63047cd6d020fc5adb186bb9cae12929735b0

Let me know what you think!

bradenmacdonald replied:
This may be the only case in the codebase where we're subclassing and override a REST API endpoint in this way though.... Either way is fine with me. Your "refer to" seems good for now, but can users actually see the other docstring in the swagger UI?

rpenido replied:
> This may be the only case in the codebase where we're subclassing and override a REST API endpoint in this way though.... Either way is fine with me. Your "refer to" seems good for now, but can users actually see the other docstring in the swagger UI?

Unfortunately no. The other API is not exposed here (only on the standalone version of openedx-learning running). A solution would be to create a static API documentation generated from Swagger, hosted on the project documentation page, but it [doesn't seems trivial](https://gist.github.com/oseiskar/dbd51a3727fc96dcf5ed189fca491fb3?permalink_comment_id=3249997).

bradenmacdonald replied:
OK, then my preference would be to use the technique I showed you to append the base class docstring to the derived class's docstring, so that it appears in the swagger docs. But any approach is fine, I don't want to block on this or spend much time on it. You can leave in "refer to" or the full copy of the docs, as you feel best.

rpenido replied:
Thank you, @bradenmacdonald! I'll take a look at this tomorrow to use the dynamic docstring.

I will also update the swagger with the API parameters. I'm using only the docstring here. Some decorators actually expose the parameters to the swagger, and we will have an even better result.

rpenido replied:
Hi @bradenmacdonald!

I started digging and found [openededx/api-doc-tools](https://github.com/openedx/api-doc-tools). I did a lot of experiments, but it didn't work well with inheritance. I also tried [drf-yasg](https://github.com/axnsan12/drf-yasg) itself. Still, I needed to put docstring in the methods (not in the view) to work with inheritance, and I ended up creating more confusion than resolving the doc string problem (it is a promising path thou, we should try it in another moment). I learned a lot but already spent too much time on this.

Do you mind reviewing the PR with the docstring in the current state?

If everything is ok, I will rebase and squash the branch.

bradenmacdonald replied:
Sounds good! We've already spent enough time on this, let's leave it there.

bradenmacdonald commented on lms/urls.py:
```diff
@@ -1049,3 +1049,8 @@
+
+# Content tagging
+urlpatterns += [
+    path('api/content_tagging/', include(('openedx.features.content_tagging.urls'))),
```
Do we need the taxonomy API in the LMS?

rpenido replied:
I'm not sure. Can you confirm @pomegranited?

pomegranited replied:
@bradenmacdonald No, we don't need the taxonomy API in the LMS.

But note that if we remove these endpoints from the LMS:

* The content_tagging APIs disappear from the swagger-generated api-docs
* The views tests need to be wrapped in [@skip_unless_cms](https://github.com/openedx/edx-platform/blob/0935e08a696204e7883f1864e7fc0f0488a3c878/openedx/core/djangolib/testing/utils.py#L244), and running them manually in the devstack requires some extra steps.

I'm fine either way.

bradenmacdonald replied:
I don't think we should allow people to make edits to content using the LMS API, and I do slightly worry that if we expose the endpoints in both, some people will use them from the LMS. But it doesn't seem like a big deal so I'm also fine either way. Having the swagger docs is a nice thing. Do the swagger docs show CMS URLs though, if you access them from the LMS? 

> running them manually in the devstack requires some extra steps.

Oh really? I haven't used devstack in a while. For tutor you always pick to run tests in LMS or CMS so there's no extra step either way.

rpenido replied:
> Do the swagger docs show CMS URLs though, if you access them from the LMS?

I don't think so. Is another port/servervice.

Tomorrow I will try to remove from LMS and see the impacts!
Thank you!

bradenmacdonald reviewed (commented):
@rpenido It says this is a draft - is this the final version?

It looks good to me. I assume @pomegranited has tested it. I just had two minor questions/comments:

bradenmacdonald commented:
@rpenido Please fix the failing test and then I can merge this.

rpenido committed:
feat: add taxonomies for org api

rpenido commented:
Hi @bradenmacdonald! Do you know how to make the tests run again here?

bradenmacdonald commented:
@rpenido They seem to be running now.

rpenido committed:
chore: run CI

rpenido commented:
I think we are good now @bradenmacdonald! Thank you!

yusuf-musleh commented:
@bradenmacdonald I just rebased my PR https://github.com/open-craft/edx-platform/pull/577 on @rpenido's work here. Was wondering if you'd like to take a take a look at it and merge it into this branch before merging this one in to master? Or are we better off doing it separately? The related openedx-learning PR https://github.com/openedx/openedx-learning/pull/68 was already merged.

bradenmacdonald commented:
@yusuf-musleh If it's just those two lines of code that were already approved, please go ahead and merge into this branch. Hope that's OK with you @rpenido.

rpenido commented:
Of course @bradenmacdonald and @yusuf-musleh! Let's reduce the overhead and earn some time!

yusuf-musleh commented:
@bradenmacdonald @rpenido Sounds good, I just need to bump up the installed version of the openedx-learning package. I'm currently running `make upgrade package=openedx-learning` locally to get the updated requirement files, will push it to my PR and merge once I confirm the tests pass (they are failing because of that),

yusuf-musleh commented:
@bradenmacdonald Had to make a small change to the on the python API side and update the relevant tests. This is after the openedx-learning version upgrade. Tests should be fixed now, waiting for them to complete, then will merge, unless you have any issues with the additional changes I added.

yusuf-musleh committed:
feat: Add retrieve object_tags REST API (#577)

rpenido committed:
chore: update requirements

rpenido committed:
Merge branch 'master' into rpenido/fal-3449-taxonomy-view-management-apis

rpenido commented on requirements/constraints.txt:
```diff
@@ -131,8 +131,3 @@ click==8.1.6
-# openedx-learning new version has some changes which are breaking quality tests
-# See https://github.com/openedx/openedx-learning/pull/68 for the changes.
-# It needs to be updated in a separate issue.
-openedx-learning==0.1.2
```
@bradenmacdonald I removed this to make the tests pass. I'm not sure what "breaking quality tests" means here.

yusuf-musleh replied:
@rpenido I believe that was because when my changes were merged to the `openedx-learning` repo, it contained changes that remove some kwargs from some functions that were being called in this repo, and that would cause the quality tests to fail (hence the version was pinned to not use the latest). Now that everything is being merged, this change should be fine, in fact, it is needed.

rpenido replied:
Thank you @yusuf-musleh!

rpenido commented:
Hi @bradenmacdonald!
Is there something we can do to help this get merged?

Thank you!

bradenmacdonald commented:
@rpenido Thanks for the ping; I actually didn't realize this was waiting for me. I'll merge it in 30 min.

openedx-webhooks commented:
@rpenido 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/32871&entry.1484655318=https://github.com/rpenido&entry.752974735=2023-07-27+20:50&entry.1917517419=2023-08-24+17:32&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
